### ISSUE TITLE
Documentation of Iterator `flatten()` improvement

### ISSUE THREAD

**Issue by ralpha** [rust-lang/rust#82687]
While linting my code I came across [`manual_flatten`](https://rust-lang.github.io/rust-clippy/master/index.html#manual_flatten) which was triggered in my code.

This suggests the change from:
```rust
for n in x {
    if let Some(n) = n {
        // something here
    }
}
```
Into:
```rust
for n in x.into_iter().flatten() {
    // something here
}
```
Both are valid code with same result. 

But the [rust docs on `flatten()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flatten) does not state that this is expected behavior. Yes, this is stated [here](https://doc.rust-lang.org/std/option/enum.Option.html#method.flatten).
But this was not strait forward (in my opinion) to find. It took me writing a whole bug report and looking a bunch of thing up to figure this out. (and I'm not even new to Rust anymore)

I would suggest adding some example code to make his explicit. Here is a suggestion:
(after the ["Mapping and then flattening: `<code>...</code>`"](https://doc.rust-lang.org/std/iter/trait.Iterator.html#examples-22) section)
> Flattening also works on other types like `Option` and `Result`:
> ```rust
> let students_in_seats = vec![Some("Hasan"), Some("Sarah"), None, Some("Tim")];
> let students_present: Vec<_> = students_in_seats.into_iter().flatten().collect();
> assert_eq!(students_present, vec!["Hasan", "Sarah", "Tim"]);
> ```
(or something similar)

This makes both this behavior cleared, easier to find and exposes people to code like this so they are less surprised if clippy starts warning them.

It looks like this behavior might be more common in the future too.
https://github.com/rust-lang/rust-clippy/issues/6061
https://github.com/rust-lang/rust-clippy/issues/6676

**Comment by cuviper** [comment]
I'm not sure I like clippy's suggestion, personally, but I think your suggested documentation is reasonable. I would also point out that _mapping_ to `Option`/`Result` followed by `flatten` (or a similar `flat_map`) is better written as `filter_map` (with `ok()` for `Result`).

**Comment by ralpha** [comment]
I was also confused when I saw the suggestion. But once I realizing what is was doing it sounded in line with things like `unwrap`. That does not mean that I like it, but at least people know it is intended that way and not a side-effect.

**Comment by nospam3089** [comment]
I encountered the same clippy warning when iterating over Results, and got perplexed. With the current state of documentation, I personally perceive the warning as suggestion to write obfuscated code hiding the fact that error checking is missing.

Adding the documentation change suggested by @ralpha would make it much easier to realize what flatten() does. Maybe it might be made even more clear how it works if phrasing the first line something like this:

> Flattening works on all types implementing Iterator, like Option and Result:

Me too am no longer new to Rust, and thought I had a proficient understanding of flatten() after having used it to unroll nested arrays. Likely I'll not be the last person to be surprised it can also be used to discard Err values.
